src/coreclr/vm/profilingenumerators.cpp

```diff
@@ -113,6 +118,10 @@ BOOL ProfilerObjectEnum::Init()
         for (unsigned segmentIdx = 0; segmentIdx < segmentsCount; segmentIdx++)
         {
             const FrozenObjectSegment* segment = segments[segmentIdx];
+            if (!segment->IsRegistered())
```

Comment:
It feels like there is a possibility for a race condition here:

1. Allocating thread left the `m_crst`, took the `m_SegmentRegistrationCrst`, called `RegisterFrozenSegment`, but having performed the volatile store yet.
2. A thread is running here, it is able to reach here because the allocating thread already left the `m_crst`

In this case, you would think the segment is not registered here but it is already registered on the allocation side.

Reply:
hm.. I don't see a good solution for this case other then using GC's lock for the whole registration part cc @jkotas 
Since profiler is expected to be executed on GC's thread.

Reply:
The earlier change run the registration part in cooperative mode. The GC cannot run when there are threads running in cooperative mode. Running the registration in cooperative mode would prevent updates while the GC is running.

When is the profiler allowed to call `GetNonGCHeapBounds`? Is it allowed to call the API only during the GC or even outside the GC?

Reply:
> When is the profiler allowed to call GetNonGCHeapBounds? Is it allowed to call the API only during the GC or even outside the GC?

A similar API the `GetNonGCHeapBounds` is based on, namely - `GetGenerationBounds` has a comment:
https://github.com/dotnet/runtime/blob/819e288da8689a5aefe9f201865b0f44f00537d4/src/coreclr/inc/corprof.idl#L2040-L2042

so I assume we're allowed to provide the same guarantee only for those callbacks too.

> The GC cannot run when there are threads running in cooperative mode. 

It seems that underlying Update/Register frozen segment can switch to preemptive, can't they? If gclock spinning takes too long it switches to preemptive and yeilds so perhaps a GC can happens once it wakes up?

Reply:
Ok, I think that we should report all frozen segments that were created by VM here, irrespective of whether they have been registered with the GC or not.

(If we do not like that behavior for some reason and we would like to only report what the GC knows about, we need to shift the responsibility to report it to the GC or get the GC to cooperate with the reporting. Otherwise, the locking schemes around this get way too complicated.)

Reply:
I prototyped a change to expose gc lock: https://github.com/EgorBo/runtime-1/commit/e5d9dacc26d7eddae8af81a5d266f29daed39e7b but overall yeah, it feels like it making NonGC heap to be rather GC heap 🙂